pkg/api/management/api/logging/validator.go

```diff
@@ -0,0 +1,87 @@
+package logging
+
+import (
+	"net"
+	"net/url"
+
+	"github.com/rancher/norman/httperror"
+	"github.com/rancher/norman/types"
+	"github.com/rancher/norman/types/convert"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+)
+
+func ClusterLoggingValidator(resquest *types.APIContext, schema *types.Schema, data map[string]interface{}) error {
+	found := false
+	if es, ok := data["ElasticsearchConfig"]; ok {
+		found = true
+		var config v3.ElasticsearchConfig
+		err := convert.ToObj(es, &config)
+		if err != nil {
+			return httperror.NewAPIError(httperror.InvalidBodyContent, "invalid field ElasticsearchConfig")
+		}
+		_, err = url.Parse(config.Endpoint)
+		if err != nil {
+			return httperror.NewAPIError(httperror.InvalidFormat, "field ElasticsearchConfig Endpoint has invalid format")
+		}
+	}
+
+	if es, ok := data["SplunkConfig"]; ok {
+		found = true
+		var config v3.SplunkConfig
+		err := convert.ToObj(es, &config)
+		if err != nil {
+			return httperror.NewAPIError(httperror.InvalidBodyContent, "invalid field SplunkConfig")
+		}
+		_, err = url.Parse(config.Endpoint)
+		if err != nil {
+			return httperror.NewAPIError(httperror.InvalidFormat, "field SplunkConfig Endpoint has invalid format")
+		}
+	}
+
+	if es, ok := data["KafkaConfig"]; ok {
+		found = true
+		var config v3.KafkaConfig
+		err := convert.ToObj(es, &config)
+		if err != nil {
+			return httperror.NewAPIError(httperror.InvalidBodyContent, "invalid field KafkaConfig")
+		}
+		if config.ZookeeperEndpoint != "" {
+			_, _, err := net.SplitHostPort(config.ZookeeperEndpoint)
+			if err != nil {
+				return httperror.NewAPIError(httperror.InvalidFormat, "field KafkaConfig ZookeeperEndpoint has invalid format")
+			}
+		}
+
+		if len(config.BrokerEndpoints) != 0 {
```

Comment:
no need to check for len being 0; the iteration on line 56 will just quit immediately if the len is 0.